src/viem/saga.ts

```diff
@@ -1,35 +1,234 @@
+ * @param options.comment the comment on the transaction
+ * @param options.feeInfo an object containing the fee information
+ * @returns
+ */
+export function* sendPayment({
+  context,
+  recipientAddress,
+  amount,
+  tokenAddress,
+  comment,
+  feeInfo,
+}: {
+  context: TransactionContext
+  recipientAddress: string
+  amount: BigNumber
+  tokenAddress: string
   comment: string
-) {
+  feeInfo: FeeInfo
+}) {
   const wallet = yield* call(getViemWallet, networkConfig.viemChain.celo)
-  Logger.debug('tokenAddress', tokenAddress, getAddress(tokenAddress))
-  const convertedAmount: string = yield* call(tokenAmountInSmallestUnit, amount, tokenAddress)
-  Logger.debug('convertedAmount', convertedAmount)
-
-  const { request } = yield* call([publicClient.celo, 'simulateContract'], {
-    address: getAddress(tokenAddress),
-    abi: erc20.abi,
-    functionName: 'transfer',
-    account: wallet.account,
-    args: [recipientAddress, convertedAmount],
+  if (!wallet.account) {
+    // this should never happen
+    throw new Error('no account found in the wallet')
+  }
+
+  Logger.debug(
+    TAG,
+    'Transferring token',
+    context.description ?? 'No description',
+    context.id,
+    tokenAddress,
+    amount,
+    feeInfo
+  )
+
+  // TODO(satish): add standby transaction
+
+  try {
+    // this returns a method which is then passed to call instead of directly
+    // doing yield* call(publicClient.celo.simulateContract, args) because this
+    // results in a long TS error
+    const simulateContractMethod = yield* call(getSimulateContractMethod, {
+      wallet,
+      tokenAddress,
+      amount,
+      recipientAddress,
+      comment,
+      feeInfo,
+    })
+
+    // Explicit type is specified here. The actual return type of this function is
+    // something like
+    // SimulateContractReturnType<typeof stableToken.abi, 'transferWithComment'>
+    // | SimulateContractReturnType<typeof erc20.abi, 'transfer'>
+    // but TSC doesn't like it when passed to writeContract
+    const { request }: SimulateContractReturnType = yield* call(simulateContractMethod)
+
+    // unlock account before executing tx
+    yield* call(unlockAccount, wallet.account.address)
+    const hash = yield* call([wallet, 'writeContract'], request)
+
+    Logger.debug(TAG, 'Transaction successfully submitted. Hash:', hash)
+
+    // TODO(satish): wait for receipt with a timeout, confirm tx and return
+
+    return hash
+  } catch (err) {
+    Logger.warn(TAG, 'Transaction failed', err)
+    throw err
+  }
+}
+
+/**
+ * Gets a function that invokes simulateContract for the appropriate contract
+ * method based on the token. If the token is a stable token, it uses the
+ * `transferWithComment` on the stable token contract, otherwise the `transfer`
+ * method on the ERC20 contract
+ *
+ * @param options an object containing the arguments
+ * @returns a function that invokes the simulateContract method
+ */
+function* getSimulateContractMethod({
```

Comment:
should this be called getSimulateTransfer? from the name alone, it sounds like it would be able to simulate other kinds of contract methods